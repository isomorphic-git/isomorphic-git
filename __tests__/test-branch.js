/* eslint-env node, browser, jasmine */
const path = require('path')
const { makeFixture } = require('./__helpers__/FixtureFS.js')
// @ts-ignore
const snapshots = require('./__snapshots__/test-branch.js.snap')
const registerSnapshots = require('./__helpers__/jasmine-snapshots')

const { branch, init, currentBranch } = require('isomorphic-git')

describe('branch', () => {
  beforeAll(() => {
    registerSnapshots(snapshots)
  })

  it('branch', async () => {
    // Setup
    const { fs, dir, gitdir } = await makeFixture('test-branch')
    // Test
    await branch({ dir, gitdir, ref: 'test-branch' })
    const files = await fs.readdir(path.resolve(gitdir, 'refs', 'heads'))
    expect(files.sort()).toMatchSnapshot()
    expect(await currentBranch({ dir, gitdir })).toEqual('master')
  })

  it('branch --checkout', async () => {
    // Setup
    const { dir, gitdir } = await makeFixture('test-branch')
    // Test
    await branch({ dir, gitdir, ref: 'test-branch', checkout: true })
    expect(await currentBranch({ dir, gitdir })).toEqual('test-branch')
  })

  it('invalid branch name', async () => {
    // Setup
    const { dir, gitdir } = await makeFixture('test-branch')
    let error = null
    // Test
    try {
      await branch({ dir, gitdir, ref: 'inv@{id..branch.lock' })
    } catch (err) {
      error = err
    }
    expect(error).not.toBeNull()
    expect(error.toJSON()).toMatchSnapshot()
  })

  it('missing ref argument', async () => {
    // Setup
    const { dir, gitdir } = await makeFixture('test-branch')
    let error = null
    // Test
    try {
      // @ts-ignore
      await branch({ dir, gitdir })
    } catch (err) {
      error = err
    }
    expect(error).not.toBeNull()
    expect(error.toJSON()).toMatchSnapshot()
  })

  it('empty repo', async () => {
    // Setup
    const { dir, fs, gitdir } = await makeFixture('test-branch-empty-repo')
    await init({ dir, gitdir })
    let error = null
    // Test
    try {
      await branch({ dir, gitdir, ref: 'test-branch', checkout: true })
    } catch (err) {
      error = err
    }
    expect(error).toBeNull()
    const file = await fs.read(path.resolve(gitdir, 'HEAD'), 'utf8')
    expect(file).toBe(`ref: refs/heads/test-branch\n`)
  })

  it('create branch with same name as a remote', async () => {
    // Setup
    const { fs, dir, gitdir } = await makeFixture('test-branch')
    let error = null
    // Test
    try {
      await branch({ dir, gitdir, ref: 'origin' })
    } catch (err) {
      error = err
    }
    expect(error).toBeNull()
    expect(
      await fs.exists(path.resolve(gitdir, 'refs/heads/origin'))
    ).toBeTruthy()
  })

  it('create branch named "HEAD"', async () => {
    // Setup
    const { fs, dir, gitdir } = await makeFixture('test-branch')
    let error = null
    // Test
    try {
      await branch({ dir, gitdir, ref: 'HEAD' })
    } catch (err) {
      error = err
    }
    expect(error).toBeNull()
    expect(
      await fs.exists(path.resolve(gitdir, 'refs/heads/HEAD'))
    ).toBeTruthy()
  })
})
