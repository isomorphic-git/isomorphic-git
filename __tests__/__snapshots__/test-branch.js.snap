// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`branch branch 1`] = `
Array [
  "master",
  "test-branch",
]
`;

exports[`branch invalid branch name 1`] = `
Object {
  "caller": "git.branch",
  "code": "InvalidRefNameError",
  "data": Object {
    "noun": "branch",
    "ref": "inv@{id..branch.lock",
    "suggestion": "inv-id.branch",
    "verb": "create",
  },
  "message": "Failed to create branch \\"inv@{id..branch.lock\\" because that name would not be a valid git reference. A valid alternative would be \\"inv-id.branch\\".",
}
`;

exports[`branch missing ref argument 1`] = `
Object {
  "caller": "git.branch",
  "code": "MissingRequiredParameterError",
  "data": Object {
    "function": "branch",
    "parameter": "ref",
  },
  "message": "The function \\"branch\\" requires a \\"ref\\" parameter but none was provided.",
}
`;
